doc/source/tune.rst

```diff
@@ -0,0 +1,160 @@
+
+The following shows grid search over two nested parameters combined with random sampling from two lambda functions. Note that the value of ``beta`` depends on the value of ``alpha``, which is represented by referencing ``spec.config.alpha`` in the lambda function. This lets you specify conditional parameter distributions.
+
+::
+
+    "config": {
+        "alpha": lambda spec: np.random.uniform(100),
+        "beta": lambda spec: spec.config.alpha * np.random.normal(),
+        "nn_layers": [
+            grid_search([16, 64, 256]),
+            grid_search([16, 64, 256]),
+        ],
+    },
+    "repeat": 10,
+
+By default, each random variable and grid search point is sampled once. To take multiple random samples or repeat grid search runs, add ``repeat: N`` to the experiment config. E.g. in the above, ``"repeat": 10`` repeats the 3x3 grid search 10 times, for a total of 90 trials, each with randomly sampled values of ``alpha`` and ``beta``.
+
+For more information on variant generation, see `variant_generator.py <https://github.com/ray-project/ray/blob/master/python/ray/tune/variant_generator.py>`__.
+
+Early Stopping
+--------------
+
+To reduce costs, long-running trials can often be early stopped if their initial performance is not promising. Ray.tune allows early stopping algorithms to be plugged in on top of existing grid or random searches. This can be enabled by setting the ``scheduler`` parameter of ``run_experiments``, e.g.
+
+::
+
+    run_experiments({...}, scheduler=MedianStoppingRule())
+
+Currently we support the following early stopping algorithms, or you can write your own that implements the `TrialScheduler <https://github.com/ray-project/ray/blob/master/python/ray/tune/trial_scheduler.py>`__ interface:
+
+.. autoclass:: ray.tune.median_stopping_rule.MedianStoppingRule
+.. autoclass:: ray.tune.hyperband.HyperBandScheduler
+
+Checkpointing support
+---------------------
+
+To enable checkpoint / resume, the full ``Trainable`` API must be implemented. This is required to support resource multiplexing in schedulers such as HyperBand. For example, all `RLlib agents <https://github.com/ray-project/ray/blob/master/python/ray/rllib/agent.py>`__ implement the ``Trainable`` API.
+
+.. autoclass:: ray.tune.trainable.Trainable
+    :members:
+
+Resource Allocation
+-------------------
+
+Ray.tune runs each trial as a Ray actor, allocating the specified GPU and CPU ``resources`` to each actor (defaulting to 1 CPU per trial). A trial will not be scheduled unless at least that amount of resources is available in the cluster, preventing the cluster from being overloaded.
+
+If your trainable function / class creates further actors that also consume CPU / GPU resources, you will also want to set ``driver_cpu_limit`` or ``driver_gpu_limit`` to tell Ray not to assign the entire resource reservation to the main Actor, as described in `trial.py <https://github.com/ray-project/ray/blob/master/python/ray/tune/trial.py>`__
```

Comment:
readers might wonder what is "the main Actor"?

Reply:
Hm how's 'top-level trainable function`?